File: src/main/java/me/relaxing9/commanditems/data/action/ActionMathExpr.java
package me.relaxing9.commanditems.data.action;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.logging.Level;

import com.fasterxml.jackson.annotation.JsonProperty;

import me.relaxing9.commanditems.CommandItems;
import me.relaxing9.commanditems.data.ItemDefinition;
import me.relaxing9.commanditems.interpreter.InterpretationContext;

public class ActionMathExpr extends Action {

    @JsonProperty(required = true)
    private String target;

    @JsonProperty(required = true)
    private String expr;

    @JsonProperty(defaultValue = "false")
    private boolean round;

    @JsonProperty(required = true)
    private Action[] actions;

    private transient Expression ast;
    private static Expression nullValue;

    public ActionMathExpr() {
        super(ActionType.MATH_EXPR);
    }

    @Override
    public void trace(List<ItemDefinition.ExecutionTrace> trace, int depth) {
        String line = String.format("%s = %s%s", target, this.round ? "(rounded) " : "", expr);

        trace.add(new ItemDefinition.ExecutionTrace(depth, line));
        for (Action action : this.actions) action.trace(trace, depth + 1);
    }

    @Override
    public void init() {
        Expression ast = parse(this.expr);
        try {
            this.ast = ast;
            nullValue = ast;

            for (Action action : this.actions) action.init();
        } catch (RuntimeException e) {
            CommandItems.logger.log(Level.SEVERE, "Failed to parse math expression: ", e);
          }
    }

    @FunctionalInterface
    public
    interface Expression {
        double eval(Map<String, Double> params);
    }

    // ====================================================
    // Derived from: https://stackoverflow.com/a/26227947
    // ====================================================
    public static Expression parse(String str) {
        return new Object() {
            int position = -1;
            int currentChar;

            void nextChar() {
                currentChar = (++position < str.length()) ? str.charAt(position) : -1;
            }

            boolean eat(int expectedChar) {
                while (currentChar == ' ') nextChar();
                if (currentChar == expectedChar) {
                    nextChar();
                    return true;
                }
                return false;
            }

            Expression parse() {
                nextChar();
                Expression expression = parseExpression();
                if (position < str.length())
                    CommandItems.logger.log(Level.WARNING, "Unexpected: " + (char) currentChar);
                return expression;
            }

            // Grammar:
            // expression = term ((ADD | SUBTRACT) term)*
            // term = factor ((MULTIPLY | DIVIDE | MODULUS) factor)*
            // factor = (ADD | SUBTRACT)? term ((EXPONENTIATE) factor)?

            Expression parseExpression() {
                Expression expression = parseTerm();
                while (true) {
                    if (eat('+')) {
                        Expression left = expression;
                        Expression right = parseTerm();
                        expression = (params) -> left.eval(params) + right.eval(params);
                    } else if (eat('-')) {
                        Expression left = expression;
                        Expression right = parseTerm();
                        expression = (params) -> left.eval(params) - right.eval(params);
                    } else return expression;
                }
            }

            Expression parseTerm() {
                Expression expression = parseFactor();
                while (true) {
                    if (eat('*')) {
                        Expression left = expression;
                        Expression right = parseFactor();
                        expression = (params) -> left.eval(params) * right.eval(params);
                    } else if (eat('/')) {
                        Expression left = expression;
                        Expression right = parseFactor();
                        expression = (params) -> left.eval(params) / right.eval(params);
                    } else return expression;
                }
            }

            Expression parseFactor() {
                if (eat('+')) return parseFactor(); // unary plus
                if (eat('-')) {
                    Expression expression = parseFactor();
                    return (params) -> -expression.eval(params); // unary minus
                }

                Expression expression = nullValue;
                int startPosition = this.position;
                if (eat('(')) { // parentheses
                    expression = parseExpression();
                    eat(')');
                } else if ((currentChar >= '0' && currentChar <= '9') || currentChar == '.') { // numbers
                    while ((currentChar >= '0' && currentChar <= '9') || currentChar == '.')
                        nextChar();
                    double result = Double.parseDouble(str.substring(startPosition, this.position));
                    expression = (params) -> result;
                } else if (Character.isLetter(currentChar) || currentChar == '_') { // symbols. May not start with a number or underscore
                    while (Character.isLetter(currentChar) || currentChar == '_' || Character.isDigit(currentChar))
                        nextChar();
                    String symbolName = str.substring(startPosition, this.position);
                    if (eat('(')) {
                        switch (symbolName) {
                            case "sqrt": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.sqrt(a.eval(params));
                                break;
                            }
                            case "sin": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.sin(a.eval(params));
                                break;
                            }
                            case "asin": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.asin(a.eval(params));
                                break;
                            }
                            case "cos": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.cos(a.eval(params));
                                break;
                            }
                            case "acos": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.acos(a.eval(params));
                                break;
                            }
                            case "tan": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.tan(a.eval(params));
                                break;
                            }
                            case "atan": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.atan(a.eval(params));
                                break;
                            }
                            case "ceil": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.ceil(a.eval(params));
                                break;
                            }
                            case "floor": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.floor(a.eval(params));
                                break;
                            }
                            case "abs": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.abs(a.eval(params));
                                break;
                            }
                            case "exp": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.exp(a.eval(params));
                                break;
                            }
                            case "log": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.log(a.eval(params));
                                break;
                            }
                            case "round": {
                                Expression a = parseExpression();
                                expression = (params) -> Math.round(a.eval(params));
                                break;
                            }
                            case "min": {
                                List<Expression> expressionList = new ArrayList<>();
                                do {
                                    Expression a = parseExpression();
                                    expressionList.add(a);
                                } while (eat(','));
                                expression = (params) -> {
                                    double min = expressionList.get(0).eval(params);
                                    for (int i = 1; i < expressionList.size(); i++) {
                                        double v = expressionList.get(i).eval(params);
                                        if (v < min)
                                            min = v;
                                    }
                                    return min;
                                };
                                break;
                            }
                            case "max": {
                                List<Expression> expressionList = new ArrayList<>();
                                do {
                                    Expression a = parseExpression();
                                    expressionList.add(a);
                                } while (eat(','));
                                expression = (params) -> {
                                    double max = expressionList.get(0).eval(params);
                                    for (int i = 1; i < expressionList.size(); i++) {
                                        double v = expressionList.get(i).eval(params);
                                        if (v > max)
                                            max = v;
                                    }
                                    return max;
                                };
                                break;
                            }
                            case "fmod": {
                                Expression a = parseExpression();
                                if (!eat(','))
                                    CommandItems.logger.log(Level.WARNING, "fmod requires two parameters!");
                                Expression b = parseExpression();

                                expression = (params) -> a.eval(params) % b.eval(params);
                                break;
                            }
                            case "sign": {
                                Expression a = parseExpression();

                                expression = (params) -> Math.signum(a.eval(params));
                                break;
                            }
                            case "rand":
                                expression = (params) -> Math.random();
                                break;

                            case "randn": {
                                Random random = new Random();
                                expression = (params) -> random.nextGaussian();
                                break;
                            }
                            default:
                                CommandItems.logger.log(Level.SEVERE, "Unknown function: " + symbolName);
                        }
                        if (!eat(')'))
                            CommandItems.logger.log(Level.WARNING, "Failed to find closing ')'.");
                    } else {
                        // Variable
                        if ("pi".equals(symbolName))
                            expression = (params) -> Math.PI;

                        else if ("e".equals(symbolName))
                            expression = (params) -> Math.E;

                        else expression = (params) -> {
                                if (!params.containsKey(symbolName))
                                    CommandItems.logger.log(Level.SEVERE, "Tried to access undefined variable: " + symbolName);
                                
                                return params.get(symbolName);
                            };
                    }
                } else {
                    CommandItems.logger.log(Level.WARNING, "Unexpected: " + (char) currentChar);
                }

                final Expression expression1 = expression;
                if (eat('^')) {
                    Expression power = parseFactor();
                    return (params) -> Math.pow(expression1.eval(params), power.eval(params)); // exponentiation
                }

                if (eat('%')) {
                    Expression modulus = parseFactor();
                    return (params) -> expression1.eval(params) % modulus.eval(params); // fmod
                }

                return expression;
            }
        }.parse();
    }
    // ====================================================

    @Override
    public void process(InterpretationContext context) {
        context.pushFrame();

        Map<String, Double> params = new HashMap<>();
        context.forEachNumericLocal(params::put);

        double rval = this.ast.eval(params);
        if (this.round)
            context.pushLocal(this.target, Long.toString(Math.round(rval)));
        else
            context.pushLocal(this.target, String.format("%f", rval));

        for (Action action : this.actions)
            action.process(context);

        context.popFrame();
    }

}

File: src/main/kotlin/me/relaxing9/commanditems/data/action/ActionMathExpr.kt
package me.relaxing9.commanditems.data.action

import com.fasterxml.jackson.annotation.JsonProperty
import me.relaxing9.commanditems.CommandItems
import me.relaxing9.commanditems.data.ItemDefinition
import me.relaxing9.commanditems.data.action.ActionMathExpr.Expression
import me.relaxing9.commanditems.interpreter.InterpretationContext
import java.util.*
import java.util.logging.Level
import kotlin.math.*

class ActionMathExpr : Action(ActionType.MATH_EXPR) {
    @JsonProperty(required = true)
    private val target: String? = null

    @JsonProperty(required = true)
    private val expr: String? = null

    @JsonProperty(defaultValue = "false")
    private val round = false

    @JsonProperty(required = true)
    private val actions: Array<Action>

    @kotlin.jvm.Transient
    private var ast: Expression? = null
    override fun trace(trace: MutableList<ItemDefinition.ExecutionTrace>, depth: Int) {
        val line: String = String.format("%s = %s%s", target, if (round) "(rounded) " else "", expr)
        trace.add(ItemDefinition.ExecutionTrace(depth, line))
        for (action in actions) action.trace(trace, depth + 1)
    }

    override fun init() {
        val ast = parse(expr)
        try {
            this.ast = ast
            nullValue = ast
            for (action in actions) action.init()
        } catch (e: RuntimeException) {
            CommandItems.logger.log(Level.SEVERE, "Failed to parse math expression: ", e)
        }
    }

    fun interface Expression {
        fun eval(params: Map<String?, Double?>?): Double
    }

    // ====================================================
    override fun process(context: InterpretationContext) {
        context.pushFrame()
        val params: MutableMap<String?, Double?> = HashMap()
        context.forEachNumericLocal { key: String?, value: Double? ->
            params[key] = value
        }
        val rval = ast!!.eval(params)
        if (round) context.pushLocal(target, rval.roundToLong().toString()) else context.pushLocal(
            target, String.format("%f", rval)
        )
        for (action in actions) action.process(context)
        context.popFrame()
    }

    companion object {
        private var nullValue: Expression? = null

        // ====================================================
        // Derived from: https://stackoverflow.com/a/26227947
        // ====================================================
        fun parse(str: String?): Expression {
            return object : Any() {
                var pos = -1
                var ch = 0
                fun nextChar() {
                    ch = if (++pos < str!!.length) str[pos].code else -1
                }

                fun eat(charToEat: Int): Boolean {
                    while (ch == ' '.code) nextChar()
                    if (ch == charToEat) {
                        nextChar()
                        return true
                    }
                    return false
                }

                fun parse(): Expression {
                    nextChar()
                    val x = parseExpression()
                    if (pos < str!!.length) CommandItems.logger.log(
                        Level.WARNING,
                        "Unexpected: " + ch.toChar()
                    )
                    return x
                }

                // Grammar:
                // expression = term | expression `+` term | expression `-` term
                // term = factor | term `*` factor | term `/` factor
                // factor = `+` factor | `-` factor | `(` expression `)`
                //        | number | functionName factor | factor `^` factor
                fun parseExpression(): Expression {
                    val x = parseTerm()
                    while (true) {
                        if (eat('+'.code)) {
                            val a: Expression = me.relaxing9.commanditems.data.action.x
                            val b = parseTerm()
                            me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? ->
                                    me.relaxing9.commanditems.data.action.a.eval(
                                        params
                                    ) + me.relaxing9.commanditems.data.action.b.eval(params)
                                } // addition
                        } else if (eat('-'.code)) {
                            val a: Expression = me.relaxing9.commanditems.data.action.x
                            val b = parseTerm()
                            me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? ->
                                    me.relaxing9.commanditems.data.action.a.eval(
                                        params
                                    ) - me.relaxing9.commanditems.data.action.b.eval(params)
                                } // subtraction
                        } else return me.relaxing9.commanditems.data.action.x
                    }
                }

                fun parseTerm(): Expression {
                    val x = parseFactor()
                    while (true) {
                        if (eat('*'.code)) {
                            val a: Expression = me.relaxing9.commanditems.data.action.x
                            val b = parseFactor()
                            me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? ->
                                    me.relaxing9.commanditems.data.action.a.eval(
                                        params
                                    ) * me.relaxing9.commanditems.data.action.b.eval(params)
                                }
                        } else if (eat('/'.code)) {
                            val a: Expression = me.relaxing9.commanditems.data.action.x
                            val b = parseFactor()
                            me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? ->
                                    me.relaxing9.commanditems.data.action.a.eval(
                                        params
                                    ) / me.relaxing9.commanditems.data.action.b.eval(params)
                                }
                        } else return me.relaxing9.commanditems.data.action.x
                    }
                }

                fun parseFactor(): Expression {
                    if (eat('+'.code)) return parseFactor() // unary plus
                    if (eat('-'.code)) {
                        val x = parseFactor()
                        return Expression { params: Map<String?, Double?>? ->
                            -me.relaxing9.commanditems.data.action.x.eval(
                                params
                            )
                        } // unary minus
                    }
                    val x = nullValue
                    val startPos = pos
                    if (eat('('.code)) { // parentheses
                        me.relaxing9.commanditems.data.action.x = parseExpression()
                        eat(')'.code)
                    } else if (ch >= '0'.code && ch <= '9'.code || ch == '.'.code) { // numbers
                        while (ch >= '0'.code && ch <= '9'.code || ch == '.'.code) nextChar()
                        val res: Double = str.substring(me.relaxing9.commanditems.data.action.startPos, pos).toDouble()
                        me.relaxing9.commanditems.data.action.x =
                            Expression { params: Map<String?, Double?>? -> me.relaxing9.commanditems.data.action.res }
                    } else if (ch >= 'a'.code && ch <= 'z'.code || ch >= 'A'.code && ch <= 'Z'.code) { // symbols. May not start with a number or underscore
                        while (ch >= 'a'.code && ch <= 'z'.code || ch >= 'A'.code && ch <= 'Z'.code || ch >= '0'.code && ch <= '9'.code || ch == '_'.code) nextChar()
                        val symbolName: String = str.substring(
                            me.relaxing9.commanditems.data.action.startPos,
                            pos
                        )
                        if (eat('('.code)) {
                            when (me.relaxing9.commanditems.data.action.symbolName) {
                                "sqrt" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            sqrt(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "sin" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            sin(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "asin" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            asin(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "cos" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            cos(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "acos" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            acos(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "tan" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            tan(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "atan" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            atan(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "ceil" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            ceil(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "floor" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            floor(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "abs" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            abs(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "exp" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            exp(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "log" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            ln(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "round" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            a.eval(params).roundToLong()
                                                .toDouble()
                                        }
                                }

                                "min" -> {
                                    val expressionList: MutableList<Expression> = ArrayList()
                                    do {
                                        val a = parseExpression()
                                        expressionList.add(a)
                                    } while (eat(','.code))
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            var min = expressionList[0].eval(params)
                                            var i = 1
                                            while (i < expressionList.size) {
                                                val v = expressionList[i].eval(params)
                                                if (v < min) min = v
                                                i++
                                            }
                                            min
                                        }
                                }

                                "max" -> {
                                    val expressionList: MutableList<Expression> = ArrayList()
                                    do {
                                        val a = parseExpression()
                                        expressionList.add(a)
                                    } while (eat(','.code))
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            var max = expressionList[0].eval(params)
                                            var i = 1
                                            while (i < expressionList.size) {
                                                val v = expressionList[i].eval(params)
                                                if (v > max) max = v
                                                i++
                                            }
                                            max
                                        }
                                }

                                "fmod" -> {
                                    val a = parseExpression()
                                    if (!eat(','.code)) CommandItems.logger.log(
                                        Level.WARNING,
                                        "fmod requires two parameters!"
                                    )
                                    val b = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            a.eval(
                                                params
                                            ) % b.eval(params)
                                        }
                                }

                                "sign" -> {
                                    val a = parseExpression()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? ->
                                            sign(
                                                a.eval(params)
                                            )
                                        }
                                }

                                "rand" -> me.relaxing9.commanditems.data.action.x =
                                    Expression { params: Map<String?, Double?>? -> Math.random() }

                                "randn" -> {
                                    val random = Random()
                                    me.relaxing9.commanditems.data.action.x =
                                        Expression { params: Map<String?, Double?>? -> random.nextGaussian() }
                                }

                                else -> CommandItems.logger.log(
                                    Level.SEVERE,
                                    "Unknown function: " + me.relaxing9.commanditems.data.action.symbolName
                                )
                            }
                            if (!eat(')'.code)) CommandItems.logger.log(Level.WARNING, "Failed to find closing ')'.")
                        } else {
                            // Variable
                            if ("pi" == me.relaxing9.commanditems.data.action.symbolName) me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? -> Math.PI } else if ("e" == me.relaxing9.commanditems.data.action.symbolName) me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?>? -> Math.E } else me.relaxing9.commanditems.data.action.x =
                                Expression { params: Map<String?, Double?> ->
                                    if (!params.containsKey(me.relaxing9.commanditems.data.action.symbolName)) CommandItems.logger.log(
                                        Level.SEVERE,
                                        "Tried to access undefined variable: " + me.relaxing9.commanditems.data.action.symbolName
                                    )
                                    params[me.relaxing9.commanditems.data.action.symbolName]!!
                                }
                        }
                    } else {
                        CommandItems.logger.log(
                            Level.WARNING,
                            "Unexpected: " + ch.toChar()
                        )
                    }
                    val x1: Expression = me.relaxing9.commanditems.data.action.x
                    if (eat('^'.code)) {
                        val p = parseFactor()
                        return Expression { params: Map<String?, Double?>? ->
                            me.relaxing9.commanditems.data.action.x1.eval(params)
                                .pow(me.relaxing9.commanditems.data.action.p.eval(params))
                        } // exponentiation
                    }
                    if (eat('%'.code)) {
                        val m = parseFactor()
                        return Expression { params: Map<String?, Double?>? ->
                            me.relaxing9.commanditems.data.action.x1.eval(
                                params
                            ) % me.relaxing9.commanditems.data.action.m.eval(params)
                        } // fmod
                    }
                    return me.relaxing9.commanditems.data.action.x
                }
            }.parse()
        }
    }
}